lib/tar.ml

```diff
@@ -524,34 +564,42 @@ module HeaderReader(Async: ASYNC)(Reader: READER with type 'a t = 'a Async.t) =
   open Reader
 
 
-  let read ?level (ifd: Reader.in_channel) : (Header.t, [ `Eof ]) result t =
+  let read ?level ?global (ifd: Reader.in_channel) : (Header.t * Header.Extended.t option, [ `Eof ]) result t =
     let level = Header.get_level level in
     (* We might need to read 2 headers at once if we encounter a Pax header *)
     let buffer = Cstruct.create Header.length in
     let real_header_buf = Cstruct.create Header.length in
 
-    let next_block () =
+    let next_block ?global () =
       really_read ifd buffer
       >>= fun () ->
-      return (Header.unmarshal ~level buffer) in
+      return (Header.unmarshal ?extended:global ~level buffer) in
 
-    (* Skip Pax GlobalExtendedHeaders *)
-    let next () =
-      next_block ()
+    let rec next ?global () =
+      next_block ?global ()
       >>= function
-      | Some x when x.Header.link_indicator = Header.Link.GlobalExtendedHeader -> next_block ()
+      | Some x when x.Header.link_indicator = Header.Link.GlobalExtendedHeader ->
+        let extra_header_buf = Cstruct.create (Int64.to_int x.Header.file_size) in
+        really_read ifd extra_header_buf
+        >>= fun () ->
+        skip ifd (Header.compute_zero_padding_length x)
+        >>= fun () ->
+        let global = Header.Extended.unmarshal ?global extra_header_buf in
```

Comment:
I wonder if we should merge with the previous `global`. I have not read up on how it works yet.

Reply:
yes.

> Each value shall affect all subsequent files that do not override that value in their own extended header record and until another global extended header record is reached that provides another value for the same field.

https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_02

Reply:
If I'm not mistaken, the code was already correct. Here, the `?global` parameter is the previous global, if present. The function `unmarshal` will merge the previous global with the current one. The test (although not super clear now that look back at it) validates the behavior.

Reply:
Done!